nft/src/main.sw

```diff
@@ -0,0 +1,474 @@
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor has already been called
+    /// - The token count is 0
+    /// - The owner is not a valid identity
+    fn constructor(owner: Identity, access_control: bool, token_supply: u64) -> bool {
+        require(storage.state == 0, InitError::CannotReinitialize);
+        require(token_supply != 0, InputError::TokenSupplyCannotBeZero);
+        require_identity_is_valid(owner);
+
+        storage.access_control_address = owner;
+        storage.access_control = access_control;
+        storage.token_supply = token_supply;
+        storage.state = 1;
+
+        true
+    }
+
+    // Uncomment when https://github.com/FuelLabs/fuels-rs/issues/375 is resolved
+    /// Returns the approved address
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The NFT contract has not been initalized
+    // fn get_approved(token_id: u64) -> Option<Identity> {
+    //     require(storage.state != 0, InitError::NFTNotInitalized);
+
+    //     let meta_data: MetaData = storage.meta_data.get(token_id);
+    //     Option::Some(meta_data.approved)
+    // }
+
+    /// Returns the tokens owned by the address
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The NFT contract has not been intialized
+    fn get_tokens(identity: Identity) -> u64 {
+        require(storage.state != 0, InitError::NFTNotInitalized);
+        storage.owners.get(identity)
+    }
+
+    /// Returns the total supply for the NFT contract
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The NFT contract has not been initalized
+    fn get_total_supply() -> u64 {
+        require(storage.state != 0, InitError::NFTNotInitalized);
+        storage.token_supply
+    }
+
+    /// Returns whether the address is approved for all tokens
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The NFT contract has not been initalized
+    fn is_approved_for_all(owner: Identity, operator: Identity) -> bool {
+        require(storage.state != 0, InitError::NFTNotInitalized);
+
+        let hash: b256 = get_owner_operator_hash(owner, operator);
+        storage.operator_approval.get(hash)
+    }
+
+    /// Mints an NFT
+    ///
+    /// # Panics
+    /// 
+    /// The function will panic when:
+    /// - The NFT contract has not been initalized 
+    /// - The amount is set to 0
+    /// - More NFTs than supply is minted
+    /// - The to Identity is not valid
+    /// - The sender is not approved to mint
+    /// - The sender sent the wrong asset
+    /// - The sender did not pay enough tokens
+    fn mint(to: Identity, amount: u64) -> bool {
+        require(storage.state != 0, InitError::NFTNotInitalized);
+        require(amount != 0, InputError::MintAmountCannotBeZero);
+        require(
+            storage.token_supply >= (storage.token_count + amount), 
```

Comment:
```suggestion
            storage.token_count + amount <= storage.token_supply, 
```

Reply:
I think you should consider if overflows are something to be considered here and if this should be rewritten. cc @nfurfaro